### PR TITLE
feat: add CFN dataset IAM role

### PR DESCRIPTION
**Pull request by bhudnell**
Issue #, if available:

Description of changes:
added functions to:
- generate a CFN IAM role with permissions necessary to perform datasets actions
- append a policy to a CFN IAM role that gives permissions to modify a single dataset

Testing was done by generating a CFN template containing the created, then updated role. The role successfully performed the dataset actions that the role has permissions for

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by SanketD92** [comment]
Could you add in the PR description the testing you performed to verify this business logic?

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    const request = {
+      roleString:
+        '{"Type":"AWS::IAM::Role","Properties":{"RoleName":"Sample-Role-Name","AssumeRolePolicyDocument":{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":"Sample-Assuming-AWS-Account-Id"},"Action":["sts:AssumeRole"]}]},"Description":"A role that allows the datasets package to perform basic actions","Policies":[{"PolicyName":"dataset-base-permissions","PolicyDocument":{"Version":"2012-10-17","Statement":[{"Sid":"AccessPointCreationDeletion","Effect":"Allow","Action":["s3:CreateAccessPoint","s3:DeleteAccessPoint","s3:GetAccessPointPolicy","s3:PutAccessPointPolicy"],"Resource":["arn:aws:s3:Sample-AWS-Bucket-Region:Sample-AWS_Account-Id:accesspoint/*"]},{"Sid":"CreateRemoveAccessPointDelegation","Effect":"Allow","Action":["s3:GetBucketPolicy","s3:PutBucketPolicy"],"Resource":["Sample-S3-Bucket-Arn"]},{"Sid":"ListTopLevelFolders","Effect":"Allow","Action":"s3:ListBucket","Resource":["Sample-S3-Bucket-Arn"],"Condition":{"StringEquals":{"s3:prefix":[""],"s3:delimiter":["/"]}}},{"Sid":"TODO","Effect":"Allow","Action":["kms:GetKeyPolicy","kms:PutKeyPolicy"],"Resource":["Sample-KMS-Key-Arn"]}]}}]}}',
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:12]: can we `JSON.stringify` this object so it's more readable?

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
...actions","Policies":[{"PolicyName":"dataset-base-permissions","PolicyDocument":{"Version":"2012-10-17","Statement":[{"Sid":"AccessPointCreationDeletion","Effect":"Allow","Action":["s3:CreateAccessPoint","s3:DeleteAccessPoint","s3:GetAccessPointPolicy","s3:PutAccessPointPolicy"],"Resource":["arn:aws:s3:Sample-AWS-Bucket-Region:Sample-AWS_Account-Id:accesspoint/*"]},{"Sid":"CreateRemoveAccessPointDelegation","Effect":"Allow","Action":["s3:GetBucketPolicy","s3:PutBucketPolicy"],"Resource":["Sample-S3-Bucket-Arn"]},{"Sid":"ListTopLevelFolders","Effect":"Allow","Action":"s3:ListBucket","Resource":["Sample-S3-Bucket-Arn"],"Condition":{"StringEquals":{"s3:prefix":[""],"s3:delimiter":["/"]}}},{"Sid":"TODO","Effect":"Allow","Action":["kms:GetKeyPolicy","kms:PutKeyPolicy"],"Resource":["Sample-KMS-Key-Arn"]}]}},{"PolicyName":"Sample-Dataset-Prefix-permissions","PolicyDocument":{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":"s3:ListBucket","Resource":["Sample-Access-Point-Arn"],"Condition":{"StringLike":{"s3:prefix":"Sample-Dataset-Prefix/*"}}},{"Effect":"Allow","Action":["s3:GetObject","s3:PutObject"],"Resource":"Sample-Access-Point-Arn/object/Sample-Dataset-Prefix/*"}]}}]}}'
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:21]: same comment as above

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    const request = {
+      roleString:
+        '{"Type":"AWS::IAM::Role","Properties":{"RoleName":"Sample-Role-Name","AssumeRolePolicyDocument":{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":"Sample-Assuming-AWS-Account-Id"},"Action":["sts:AssumeRole"]}]},"Description":"A role that allows the datasets package to perform basic actions","Policies":[{"PolicyName":"dataset-base-permissions","PolicyDocument":{"Version":"2012-10-17","Statement":[{"Sid":"AccessPointCreationDeletion","Effect":"Allow","Action":["s3:CreateAccessPoint","s3:DeleteAccessPoint","s3:GetAccessPointPolicy","s3:PutAccessPointPolicy"],"Resource":["arn:aws:s3:Sample-AWS-Bucket-Region:Sample-AWS_Account-Id:accesspoint/*"]},{"Sid":"CreateRemoveAccessPointDelegation","Effect":"Allow","Action":["s3:GetBucketPolicy","s3:PutBucketPolicy"],"Resource":["Sample-S3-Bucket-Arn"]},{"Sid":"ListTopLevelFolders","Effect":"Allow","Action":"s3:ListBucket","Resource":["Sample-S3-Bucket-Arn"],"Condition":{"StringEquals":{"s3:prefix":[""],"s3:delimiter":["/"]}}},{"Sid":"TODO","Effect":"Allow","Action":["kms:GetKeyPolicy","kms:PutKeyPolicy"],"Resource":["Sample-KMS-Key-Arn"]}]}}]}}',
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:28]: same comment as above

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      accessPointArn: 'Sample-Access-Point-Arn',
+      datasetPrefix: 'Sample-Dataset-Prefix'
+    };
+
+    expect(() => {
+      addDatasetPermissionsToRole({ ...request, roleString: '' });
+    }).toThrow(Error); // TODO replace with custom error
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:35]: where is this error defined?

File: workbench-core/datasets/src/awsUtilities/generateBaseDatasetRole.test.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    const response = generateBaseDatasetsRole(baseRequest);
+
+    expect(response).toStrictEqual({
+      iamRoleString:
+        '{"Type":"AWS::IAM::Role","Properties":{"RoleName":"Sample-Role-Name","AssumeRolePolicyDocument":{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":"Sample-Assuming-AWS-Account-Id"},"Action":["sts:AssumeRole"]}]},"Description":"A role that allows the datasets package to perform basic actions","Policies":[{"PolicyName":"dataset-base-permissions","PolicyDocument":{"Version":"2012-10-17","Statement":[{"Sid":"AccessPointCreationDeletion","Effect":"Allow","Action":["s3:CreateAccessPoint","s3:DeleteAccessPoint","s3:GetAccessPointPolicy","s3:PutAccessPointPolicy"],"Resource":["arn:aws:s3:Sample-AWS-Bucket-Region:Sample-AWS_Account-Id:accesspoint/*"]},{"Sid":"CreateRemoveAccessPointDelegation","Effect":"Allow","Action":["s3:GetBucketPolicy","s3:PutBucketPolicy"],"Resource":["Sample-S3-Bucket-Arn"]},{"Sid":"ListTopLevelFolders","Effect":"Allow","Action":"s3:ListBucket","Resource":["Sample-S3-Bucket-Arn"],"Condition":{"StringEquals":{"s3:prefix":[""],"s3:delimiter":["/"]}}}]}}]}}'
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/generateBaseDatasetRole.test.ts:22]: same comment as above

File: workbench-core/datasets/src/awsUtilities/generateBaseDatasetRole.test.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    expect(response).toStrictEqual({
+      iamRoleString:
+        '{"Type":"AWS::IAM::Role","Properties":{"RoleName":"Sample-Role-Name","AssumeRolePolicyDocument":{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":"Sample-Assuming-AWS-Account-Id"},"Action":["sts:AssumeRole"]}]},"Description":"A role that allows the datasets package to perform basic actions","Policies":[{"PolicyName":"dataset-base-permissions","PolicyDocument":{"Version":"2012-10-17","Statement":[{"Sid":"AccessPointCreationDeletion","Effect":"Allow","Action":["s3:CreateAccessPoint","s3:DeleteAccessPoint","s3:GetAccessPointPolicy","s3:PutAccessPointPolicy"],"Resource":["arn:aws:s3:Sample-AWS-Bucket-Region:Sample-AWS_Account-Id:accesspoint/*"]},{"Sid":"CreateRemoveAccessPointDelegation","Effect":"Allow","Action":["s3:GetBucketPolicy","s3:PutBucketPolicy"],"Resource":["Sample-S3-Bucket-Arn"]},{"Sid":"ListTopLevelFolders","Effect":"Allow","Action":"s3:ListBucket","Resource":["Sample-S3-Bucket-Arn"],"Condition":{"StringEquals":{"s3:prefix":[""],"s3:delimiter":["/"]}}},{"Sid":"TODO","Effect":"Allow","Action":["kms:GetKeyPolicy","kms:PutKeyPolicy"],"Resource":["Sample-KMS-Key-Arn"]}]}}]}}'
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/generateBaseDatasetRole.test.ts:36]: same comment

File: workbench-core/datasets/src/awsUtilities/generateBaseDatasetsRole.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  s3BucketArn: string;
+  assumingAwsAccountId: string;
+  kmsKeyArn?: string;
+}
+
+export interface GenerateBaseDatasetsRoleResponse {
+  iamRoleString: string;
+}
+
+export function generateBaseDatasetsRole(
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/generateBaseDatasetsRole.ts:19]: Add tsdoc comments

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  roleString: string;
+  accessPointArn: string;
+  datasetPrefix: string;
+}
+
+export interface GenerateDatasetPermissionsResponse {
+  iamRoleString: string;
+}
+
+export function addDatasetPermissionsToRole(
```
Comment (SanketD92) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.ts:18]: add tsdoc comments

File: workbench-core/datasets/src/awsUtilities/generateBaseDatasetsRole.ts (hunk 7 above)
Comment (SanketD92) [reply workbench-core/datasets/src/awsUtilities/generateBaseDatasetsRole.ts:19]: suggested method rename: `registerExternalBucket`

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts (hunk 1 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:12]: done

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts (hunk 2 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:21]: done

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts (hunk 3 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:28]: refactored, so not necessary

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts (hunk 4 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:35]: fixed

File: workbench-core/datasets/src/awsUtilities/generateBaseDatasetsRole.ts (hunk 7 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/generateBaseDatasetsRole.ts:19]: done

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.ts (hunk 8 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.ts:18]: done

**Review by SanketD92 — APPROVED** [review]

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    };
+
+    const response = addDatasetPermissionsToRole(request);
+
+    expect(response).toStrictEqual({
+      iamRoleString: JSON.stringify(responseRole)
+    });
+  });
+
+  it('throws an InvalidRoleStringError when the roleString is malformed', () => {
```
Comment (htztomic) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:175]:
```suggestion
  it('throws an InvalidIamRoleError when the roleString is malformed', () => {
```

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    const request = {
+      roleString: '',
+      accessPointArn: 'Sample-Access-Point-Arn',
+      datasetPrefix: 'Sample-Dataset-Prefix'
+    };
+
+    try {
+      addDatasetPermissionsToRole(request);
```
Comment (htztomic) [inline workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:183]: add expect.hasAssertions()

File: workbench-core/datasets/src/awsUtilities/createRegisterExternalBucketRole.test.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    });
+  });
+
+  it('includes KMS key policy when a KMS key Arn is included', () => {
+    const request = {
+      ...baseRequest,
+      kmsKeyArn: 'Sample-KMS-Key-Arn'
+    };
+    const roleString = JSON.stringify({
+      Type: 'AWS::IAM::Role',
```
Comment (htztomic) [inline workbench-core/datasets/src/awsUtilities/createRegisterExternalBucketRole.test.ts:90]: can you extract this to beforeEach and just modify it

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * The updated JSON.stringified IAM role
+   */
+  iamRoleString: string;
+}
+
+/**
+ * Add a policy to an existing IAM role with permissions to modify the defined dataset
+ * @param request - {@link AddDatasetPermissionsToRoleRequest}
+ * @returns - {@link AddDatasetPermissionsToRoleResponse}
```
[the remark this answers is not in the document]
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.ts:36]: done

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts (hunk 9 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:175]: done

File: workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts (hunk 10 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/addDatasetPermissionsToRole.test.ts:183]: done

File: workbench-core/datasets/src/awsUtilities/createRegisterExternalBucketRole.test.ts (hunk 11 above)
Comment (bhudnell) [reply workbench-core/datasets/src/awsUtilities/createRegisterExternalBucketRole.test.ts:90]: done

**Review by SanketD92 — APPROVED** [review]

**Review by htztomic — APPROVED** [review]
LGTM

**Pull request merged**